Do some more std::span adoption in WebCore

cdumez opened the pull request:
#### fa32236821c30e4b79550d58ae4e12e3e7aabe5d
<pre>
Do some more std::span adoption in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=272437">https://bugs.webkit.org/show_bug.cgi?id=272437</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/SortedArrayMap.h:
(WTF::lessThanASCIICaseFolding):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::copyBaseWindowsDriveLetter):
(WTF::URLParser::shouldPopPath):
(WTF::URLParser::isLocalhost):
(WTF::URLParser::URLParser):
(WTF::URLParser::parse):
(WTF::URLParser::subdomainStartsWithXNDashDash):
(WTF::URLParser::parseHostAndPort):
* Source/WTF/wtf/URLParser.h:
* Source/WTF/wtf/text/CodePointIterator.h:
(WTF::CodePointIterator::CodePointIterator):
(WTF::CodePointIterator::operator==):
(WTF::CodePointIterator::atEnd const):
(WTF::CodePointIterator::codeUnitsSince const):
(WTF::CodePointIterator&lt;LChar&gt;::operator const):
(WTF::CodePointIterator&lt;LChar&gt;::operator):
(WTF::CodePointIterator&lt;UChar&gt;::operator const):
(WTF::CodePointIterator&lt;UChar&gt;::operator):
(): Deleted.
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::eucJPEncode):
(PAL::iso2022JPEncode):
(PAL::shiftJISEncode):
(PAL::eucKREncode):
(PAL::big5Encode):
(PAL::gbEncodeShared):

Canonical link: <a href="https://commits.webkit.org/277448@main">https://commits.webkit.org/277448@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82beadf35d9cbe805ae06ad78cd990b054f2b2c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47585 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26768 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50249 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50267 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43633 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49892 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32521 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24229 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38742 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48167 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24384 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41005 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20045 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21869 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42180 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5627 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/40866 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC (warnings); Running jscore-test") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43914 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42601 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52148 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47075 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22619 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18955 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46045 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Skipped layout-tests; 110 flakes 48 failures; Uploaded test results; 12 flakes 31 failures; Running compile-webkit-without-change") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23892 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45076 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24681 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54575 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6728 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23611 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11187 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -2703,12 +2703,9 @@ bool URLParser::subdomainStartsWithXNDashDash(CodePointIterator<CharacterType> i
-    const UChar* begin = host.characters16();
-    return subdomainStartsWithXNDashDash(CodePointIterator<UChar>(begin, begin + host.length()));
+    if (host.is8Bit())
+        return subdomainStartsWithXNDashDash(CodePointIterator<LChar>(host.span8()));
```
Can we rearrange things so we don’t have to write the class name `CodePointIterator` out here? Maybe if we add a deduction guide we can implicitly convert in most places like this.

darinadler commented on Source/WTF/wtf/text/CodePointIterator.h:
```diff
@@ -94,17 +94,17 @@ ALWAYS_INLINE char32_t CodePointIterator<UChar>::operator*() const
     ASSERT(!atEnd());
     char32_t c;
-    U16_GET(m_begin, 0, 0, m_end - m_begin, c);
+    U16_GET(m_data.data(), 0, 0, m_data.size(), c);
```
Should just use `m_data` here instead of `m_data.data()`.

darinadler commented on Source/WTF/wtf/text/CodePointIterator.h:
```diff
@@ -35,57 +35,57 @@ template<typename CharacterType>
     
     ALWAYS_INLINE CodePointIterator(const CodePointIterator& begin, const CodePointIterator& end)
-        : CodePointIterator(begin.m_begin, end.m_begin)
+        : CodePointIterator({ begin.m_data.data(), end.m_data.data() })
```
Can’t this just be `begin.m_data` without the braces?

cdumez replied:
I don't think so, this is essentially merging the spans from `begin` and `end` CodePointIterators. We can't ignore `end`.

darinadler replied:
Oops, didn’t notice that!

darinadler commented on Source/WTF/wtf/text/CodePointIterator.h:
```diff
@@ -35,57 +35,57 @@ template<typename CharacterType>
-    ALWAYS_INLINE friend bool operator==(const CodePointIterator&, const CodePointIterator&) = default;
+    ALWAYS_INLINE friend bool operator==(const CodePointIterator& a, const CodePointIterator& b)
+    {
+        return a.m_data.data() == b.m_data.data() && a.m_data.size() == b.m_data.size();
```
Can’t we use the `std::span` version of `==` here? I’m surprised we can’t just use default like before.

cdumez replied:
ahah, once would indeed assume that std::span has a comparison operator, and yet...
The standard currently doesn't have one, though I believe there is a proposal to add one.

darinadler replied:
I see

darinadler commented on Source/WTF/wtf/text/CodePointIterator.h:
```diff
@@ -35,57 +35,57 @@ template<typename CharacterType>
-        ASSERT(m_begin >= reference);
-        return m_begin - reference;
+        ASSERT(m_data.data() >= reference);
+        return m_data.data() - reference;
```
Not sure when to use pointers and when to use `std::span` iterators instead.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
```diff
@@ -273,7 +273,7 @@ static Vector<uint8_t> eucJPEncode(StringView string, Function<void(char32_t, Ve
 
     auto characters = string.upconvertedCharacters();
-    for (WTF::CodePointIterator<UChar> iterator(characters.get(), characters.get() + string.length()); !iterator.atEnd(); ++iterator) {
+    for (WTF::CodePointIterator<UChar> iterator({ characters.get(), string.length() }); !iterator.atEnd(); ++iterator) {
```
Should `upconvertedCharacters` turn itself into a `span` instead of a pointer or in addition to a pointer?

cdumez replied:
Yes, I think we need to update upconvertedCharacters. It currently only has a pointer and doesn't know about its length. The length has to come from somewhere else (usually the original string), which is too error prone. I have an item on my TODO list to look into it.

darinadler replied:
Since it’s only used in cases that we don’t want to optimize for performance it should be an easy change. It’s intended to be a convenience to let you handle cases where we always want to use `char16_t` without having to write out buffer management code. The reason it only produced a pointer was the kind of idiom it was designed for. You’d only call it right where you’d have to pass a pointer to something that needs UTF-16. The original idea was that if the code is performance critical we can always use more efficient idioms that are more verbose.

cdumez replied:
Ok, I have a patch ready for upconvertedCharacters but I'll perf A/B test it just in case.

cdumez committed:
Do some more std::span adoption in WebCore
https://bugs.webkit.org/show_bug.cgi?id=272437

Reviewed by Darin Adler.

* Source/WTF/wtf/SortedArrayMap.h:
(WTF::lessThanASCIICaseFolding):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::copyBaseWindowsDriveLetter):
(WTF::URLParser::shouldPopPath):
(WTF::URLParser::isLocalhost):
(WTF::URLParser::URLParser):
(WTF::URLParser::parse):
(WTF::URLParser::subdomainStartsWithXNDashDash):
(WTF::URLParser::parseHostAndPort):
* Source/WTF/wtf/URLParser.h:
* Source/WTF/wtf/text/CodePointIterator.h:
(WTF::CodePointIterator::CodePointIterator):
(WTF::CodePointIterator::operator==):
(WTF::CodePointIterator::atEnd const):
(WTF::CodePointIterator::codeUnitsSince const):
(WTF::CodePointIterator<LChar>::operator const):
(WTF::CodePointIterator<LChar>::operator):
(WTF::CodePointIterator<UChar>::operator const):
(WTF::CodePointIterator<UChar>::operator):
(): Deleted.
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::eucJPEncode):
(PAL::iso2022JPEncode):
(PAL::shiftJISEncode):
(PAL::eucKREncode):
(PAL::big5Encode):
(PAL::gbEncodeShared):

Canonical link: https://commits.webkit.org/277448@main
